File: ml-benchmark/fusion/algprog/ex3.13.f
Config NonLinear = true;

Inductive List = nil Unit | cons {Int, List};
Inductive PList = pnil Unit | pcons {Int, Int, PList};

tri = fix (
  \f: List -> Compress PList. \xs: List.
  match xs with
    nil _ -> pnil unit
  | cons {h, t} ->
    let tail = (fix (
      \g: PList -> PList. \ys: PList.
      match ys with
        pnil _ -> pnil unit
      | pcons {n, h, t} -> pcons {+ n 1, h, g t}
      end
    )) (f t) in pcons {0, h, tail}
  end
);

tsum = fix (
  \f: PList -> Int. \xs: PList.
  match xs with
    pnil _ -> 0
  | pcons {n, h, t} -> + (* n h) (f t)
  end
);

main = \xs: List. tsum (tri xs);

File: ml-benchmark/fusion/algprog/ex3.13.ml
type list_ = Nil | Cons of int * list_
type plist = PNil | PCons of int * int * plist

let rec tri xs =
  match xs with
  | Nil -> PNil
  | Cons (h, t) ->
    let tail =
      (let rec g ys =
        match ys with
        | PNil -> PNil
        | PCons (n, h2, t2) -> PCons (n + 1, h2, g t2)
      in g) (tri t)
    in
    PCons (0, h, tail)

let rec tsum xs =
  match xs with
  | PNil -> 0
  | PCons (n, h, t) -> n * h + tsum t

let program xs = tsum (tri xs)